third_party/move/move-compiler-v2/src/pipeline/uninitialized_use_checker.rs

```diff
@@ -0,0 +1,386 @@
+    /// The function's `num_params` and `num_locals` are provided.
+    ///
+    /// Note: `num_locals` is the total number of locals, including params.
+    /// Thus, `num_locals` must be >= `num_params`.
+    pub fn new(num_params: usize, num_locals: usize) -> Self {
+        Self {
+            num_params,
+            num_locals,
+        }
+    }
+
+    /// Analyze the given function and return the initialized state of all locals before and
+    /// after each reachable program point.
+    fn analyze(&self, func_target: &FunctionTarget) -> InitializedStateAnnotation {
+        let code = func_target.get_bytecode();
+        let cfg = StacklessControlFlowGraph::new_forward(code);
+        let block_state_map = self.analyze_function(
+            InitializedState::new(self.num_params, self.num_locals),
+            code,
+            &cfg,
+        );
+        let per_bytecode_state =
+            self.state_per_instruction(block_state_map, code, &cfg, |before, after| {
+                (before.clone(), after.clone())
+            });
+        InitializedStateAnnotation(per_bytecode_state)
+    }
+}
+
+impl TransferFunctions for InitializedStateAnalysis {
+    type State = InitializedState;
+
+    // This is a forward analysis.
+    const BACKWARD: bool = false;
+
+    fn execute(&self, state: &mut Self::State, instr: &Bytecode, _offset: CodeOffset) {
+        // Once you write to a local, it is considered initialized.
+        instr.dests().iter().for_each(|dst| {
+            state.mark_as_initialized(*dst);
+        });
+    }
+}
+
+impl DataflowAnalysis for InitializedStateAnalysis {}
+
+/// Checker which verifies that all locals are definitely initialized before use.
+/// Violations are reported as errors.
+pub struct UninitializedUseChecker {}
+
+impl UninitializedUseChecker {
+    /// Check whether all locals are definitely initialized before use in the function `target`.
+    /// Information about initialized state is provided in `annotation`.
+    /// Violations are reported as errors in the `target`'s global environment.
+    fn perform_checks(&self, target: &FunctionTarget, annotation: &InitializedStateAnnotation) {
+        for (offset, bc) in target.get_bytecode().iter().enumerate() {
+            if bc.is_spec_only() {
+                // We don't check spec-only instructions here.
+                continue;
+            }
+            bc.sources().iter().for_each(|src| {
+                if let Some(state @ (Initialized::Maybe | Initialized::No)) =
+                    annotation.get_initialized_state(*src, offset as CodeOffset)
+                {
+                    target.global_env().error(
+                        &target.get_bytecode_loc(bc.get_attr_id()),
+                        &format!(
+                            "use of {}unassigned {}",
+                            match state {
+                                Initialized::No => "",
+                                Initialized::Maybe => "possibly ",
```

Comment:
You could reduce the code to understand/test by combining the last 2 cases as _.

Reply:
Agreed.